File: p5/cpu_scheduling.py
# CPU Scheduling Simulator in Python
# Algorithms: FCFS, SJF (Preemptive), Priority (Non-Preemptive), Round Robin (Preemptive)

def find_waiting_time_fcfs(processes):
    wt = [0] * len(processes)
    processes.sort(key=lambda x: x[1])  # Sort by arrival time
    service_time = processes[0][1]
    for i in range(1, len(processes)):
        service_time += processes[i-1][2]
        wt[i] = max(0, service_time - processes[i][1])
    return wt

def find_turnaround_time(processes, wt):
    tat = [processes[i][2] + wt[i] for i in range(len(processes))]
    return tat

def find_avg_time_fcfs(processes):
    wt = find_waiting_time_fcfs(processes)
    tat = find_turnaround_time(processes, wt)
    print("\nFCFS Scheduling:")
    print("PID\tAT\tBT\tWT\tTAT")
    for i in range(len(processes)):
        print(f"{processes[i][0]}\t{processes[i][1]}\t{processes[i][2]}\t{wt[i]}\t{tat[i]}")
    print(f"Average Waiting Time: {sum(wt)/len(wt):.2f}")
    print(f"Average Turnaround Time: {sum(tat)/len(tat):.2f}")

# ---------------- SJF Preemptive ----------------
def sjf_preemptive(processes):
    n = len(processes)
    rt = [p[2] for p in processes]
    complete = 0
    t = 0
    wt = [0]*n
    tat = [0]*n
    minm = 999999999
    short = 0
    check = False

    while complete != n:
        for j in range(n):
            if processes[j][1] <= t and rt[j] < minm and rt[j] > 0:
                minm = rt[j]
                short = j
                check = True
        if not check:
            t += 1
            continue

        rt[short] -= 1
        minm = rt[short] if rt[short] > 0 else 999999999

        if rt[short] == 0:
            complete += 1
            check = False
            fint = t + 1
            wt[short] = (fint - processes[short][2] - processes[short][1])
            if wt[short] < 0:
                wt[short] = 0
        t += 1

    tat = [processes[i][2] + wt[i] for i in range(n)]
    print("\nSJF (Preemptive):")
    print("PID\tAT\tBT\tWT\tTAT")
    for i in range(n):
        print(f"{processes[i][0]}\t{processes[i][1]}\t{processes[i][2]}\t{wt[i]}\t{tat[i]}")
    print(f"Average Waiting Time: {sum(wt)/n:.2f}")
    print(f"Average Turnaround Time: {sum(tat)/n:.2f}")

# ---------------- Priority Non-preemptive ----------------
def priority_non_preemptive(processes):
    processes.sort(key=lambda x: (x[1], x[3]))  # Sort by arrival then priority
    wt = [0]*len(processes)
    total_bt = 0
    for i in range(1, len(processes)):
        total_bt += processes[i-1][2]
        wt[i] = max(0, total_bt - processes[i][1])
    tat = [processes[i][2] + wt[i] for i in range(len(processes))]

    print("\nPriority Scheduling (Non-preemptive):")
    print("PID\tAT\tBT\tPR\tWT\tTAT")
    for i in range(len(processes)):
        print(f"{processes[i][0]}\t{processes[i][1]}\t{processes[i][2]}\t{processes[i][3]}\t{wt[i]}\t{tat[i]}")
    print(f"Average Waiting Time: {sum(wt)/len(wt):.2f}")
    print(f"Average Turnaround Time: {sum(tat)/len(tat):.2f}")

# ---------------- Round Robin Preemptive ----------------
def round_robin(processes, quantum):
    n = len(processes)
    rem_bt = [p[2] for p in processes]
    t = 0
    wt = [0]*n

    while True:
        done = True
        for i in range(n):
            if rem_bt[i] > 0:
                done = False
                if rem_bt[i] > quantum:
                    t += quantum
                    rem_bt[i] -= quantum
                else:
                    t += rem_bt[i]
                    wt[i] = t - processes[i][2] - processes[i][1]
                    rem_bt[i] = 0
        if done:
            break
    tat = [processes[i][2] + wt[i] for i in range(n)]

    print("\nRound Robin (Preemptive):")
    print("PID\tAT\tBT\tWT\tTAT")
    for i in range(n):
        print(f"{processes[i][0]}\t{processes[i][1]}\t{processes[i][2]}\t{wt[i]}\t{tat[i]}")
    print(f"Average Waiting Time: {sum(wt)/n:.2f}")
    print(f"Average Turnaround Time: {sum(tat)/n:.2f}")

# ---------------- Main ----------------
if __name__ == "__main__":
    processes = [
        [1, 0, 5, 2],   # [PID, Arrival Time, Burst Time, Priority]
        [2, 1, 3, 1],
        [3, 2, 8, 3],
        [4, 3, 6, 2]
    ]

    find_avg_time_fcfs(processes.copy())
    sjf_preemptive(processes.copy())
    priority_non_preemptive(processes.copy())
    round_robin(processes.copy(), quantum=2)

File: p5/cpu_scheduling.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

struct Process {
    int pid;
    int arrival;
    int burst;
    int priority;
    int remaining;
    int waiting;
    int turnaround;
};

// ---------- FCFS ----------
void fcfs(vector<Process> p) {
    cout << "\n=== FCFS Scheduling ===\n";
    int time = 0;
    for (auto &pr : p) {
        if (time < pr.arrival) time = pr.arrival;
        pr.waiting = time - pr.arrival;
        time += pr.burst;
        pr.turnaround = pr.waiting + pr.burst;
        cout << "P" << pr.pid << " ";
    }

    double avgWait = 0, avgTurn = 0;
    for (auto &pr : p) {
        avgWait += pr.waiting;
        avgTurn += pr.turnaround;
    }

    cout << "\nAverage Waiting Time: " << avgWait / p.size();
    cout << "\nAverage Turnaround Time: " << avgTurn / p.size() << "\n";
}

// ---------- SJF (Preemptive) ----------
void sjfPreemptive(vector<Process> p) {
    cout << "\n=== SJF (Preemptive) Scheduling ===\n";
    int completed = 0, time = 0, n = p.size();
    int shortest = -1, minBurst = 1e9;
    for (auto &pr : p) pr.remaining = pr.burst;

    while (completed < n) {
        shortest = -1;
        minBurst = 1e9;

        for (int i = 0; i < n; i++) {
            if (p[i].arrival <= time && p[i].remaining > 0 && p[i].remaining < minBurst) {
                minBurst = p[i].remaining;
                shortest = i;
            }
        }

        if (shortest == -1) { time++; continue; }

        p[shortest].remaining--;
        cout << "P" << p[shortest].pid << " ";
        time++;

        if (p[shortest].remaining == 0) {
            completed++;
            p[shortest].turnaround = time - p[shortest].arrival;
            p[shortest].waiting = p[shortest].turnaround - p[shortest].burst;
        }
    }

    double avgWait = 0, avgTurn = 0;
    for (auto &pr : p) {
        avgWait += pr.waiting;
        avgTurn += pr.turnaround;
    }

    cout << "\nAverage Waiting Time: " << avgWait / n;
    cout << "\nAverage Turnaround Time: " << avgTurn / n << "\n";
}

// ---------- Priority (Non-Preemptive) ----------
void priorityScheduling(vector<Process> p) {
    cout << "\n=== Priority (Non-Preemptive) Scheduling ===\n";
    int time = 0, completed = 0;
    int n = p.size();
    vector<int> done(n, 0);

    while (completed < n) {
        int idx = -1, bestPri = 1e9;
        for (int i = 0; i < n; i++) {
            if (p[i].arrival <= time && !done[i] && p[i].priority < bestPri) {
                bestPri = p[i].priority;
                idx = i;
            }
        }

        if (idx == -1) { time++; continue; }

        p[idx].waiting = time - p[idx].arrival;
        time += p[idx].burst;
        p[idx].turnaround = p[idx].waiting + p[idx].burst;
        done[idx] = 1;
        completed++;
        cout << "P" << p[idx].pid << " ";
    }

    double avgWait = 0, avgTurn = 0;
    for (auto &pr : p) {
        avgWait += pr.waiting;
        avgTurn += pr.turnaround;
    }

    cout << "\nAverage Waiting Time: " << avgWait / n;
    cout << "\nAverage Turnaround Time: " << avgTurn / n << "\n";
}

// ---------- Round Robin ----------
void roundRobin(vector<Process> p, int quantum) {
    cout << "\n=== Round Robin Scheduling ===\n";
    queue<int> q;
    int time = 0, n = p.size(), completed = 0;

    for (auto &pr : p) pr.remaining = pr.burst;
    vector<int> inQueue(n, 0);

    while (completed < n) {
        for (int i = 0; i < n; i++)
            if (p[i].arrival <= time && !inQueue[i] && p[i].remaining > 0) {
                q.push(i);
                inQueue[i] = 1;
            }

        if (q.empty()) { time++; continue; }

        int idx = q.front(); q.pop();
        inQueue[idx] = 0;

        int exec = min(quantum, p[idx].remaining);
        p[idx].remaining -= exec;
        for (int t = 0; t < exec; t++)
            cout << "P" << p[idx].pid << " ";
        time += exec;

        if (p[idx].remaining == 0) {
            completed++;
            p[idx].turnaround = time - p[idx].arrival;
            p[idx].waiting = p[idx].turnaround - p[idx].burst;
        } else {
            for (int i = 0; i < n; i++)
                if (p[i].arrival <= time && !inQueue[i] && p[i].remaining > 0) {
                    q.push(i);
                    inQueue[i] = 1;
                }
            q.push(idx);
            inQueue[idx] = 1;
        }
    }

    double avgWait = 0, avgTurn = 0;
    for (auto &pr : p) {
        avgWait += pr.waiting;
        avgTurn += pr.turnaround;
    }

    cout << "\nAverage Waiting Time: " << avgWait / n;
    cout << "\nAverage Turnaround Time: " << avgTurn / n << "\n";
}

// ---------- MAIN ----------
int main() {
    int n;
    cout << "Enter number of processes: ";
    cin >> n;

    vector<Process> p(n);
    for (int i = 0; i < n; i++) {
        p[i].pid = i + 1;
        cout << "Enter Arrival, Burst, Priority for P" << i + 1 << ": ";
        cin >> p[i].arrival >> p[i].burst >> p[i].priority;
    }

    // Sort by arrival for fair start
    sort(p.begin(), p.end(), [](Process a, Process b) { return a.arrival < b.arrival; });

    fcfs(p);
    sjfPreemptive(p);
    priorityScheduling(p);
    roundRobin(p, 2);  // quantum = 2

    return 0;
}
